### PR TITLE
Rolling Upgrade: Support master node type change

### LINKED ISSUE #1710: Rolling upgrade : update predicate to handle "master+data" to "data only" (and  vice versa)
```
11:11:47      --- FAIL: TestRiskyMasterReconfiguration/Verify_that_a_config_change_is_being_applied (300.00s)
11:11:47          require.go:794: 
11:11:47                  Error Trace:    utils.go:80
11:11:47                  Error:          Received unexpected error:
11:11:47                                  no upgrade in progress
11:11:47                  Test:           TestRiskyMasterReconfiguration/Verify_that_a_config_change_is_being_applied
```

There're 2 bugs here:

1. [`StatefulSetList.ToUpdate`](https://github.com/elastic/cloud-on-k8s/blob/master/pkg/controller/elasticsearch/sset/list.go#L53-L62) does not work as expected since we move to `OnDelete`. This is because the status of the `currentRevision` is not updated anymore.

2. A master node can't be downgraded to a data node because a predicate wrongly assume that there are more healthy masters than expected. In the example above there are 2 healthy masters but only one is expected.

First bug should be easy to fix as we should not rely anymore on the `currentRevision` in the statefulset status. We should compare `updateRevision` with the revision stored in the annotation on the Pods.

Second one is a bit more tricky, here is a proposal:

* Every mutation on a master node should be tracked with a boolean to ensure that we only update 1 and only 1 master at a time [similarly to what is done for downscale](https://github.com/elastic/cloud-on-k8s/blob/master/pkg/controller/elasticsearch/driver/downscale_invariants.go#L40)
* When a "master+data node" is updated into a "data node only" I think we can reuse the logic in [`checkDownscaleInvariant`](https://github.com/elastic/cloud-on-k8s/blob/master/pkg/controller/elasticsearch/driver/downscale_invariants.go#L23)
* When a "data node only" is upgraded to a "master+data" node I'm not sure that there is anything else to do that to ensure that it is done one by one.

I was expecting to work on these issues with #1693  but it looks like I have to work on this one before.

### PR DESCRIPTION
**Pull request by barkbay**
This PR adds some support for the master node type change.
It's a partial fix for #1693: only master changes are handled and tested in this PR.
I will open two new issues:

1. one for data node changes: MD &rightarrow; M and M &rightarrow; MD
2. one to discuss type inversions: M &rightarrow; D or D &rightarrow; M

Also fix #1710 and #1628

### CODE REVIEW

File: pkg/controller/elasticsearch/driver/upgrade_pods_deletion.go [hunk 1]
```diff
@@ -120,6 +124,30 @@ func sortCandidates(allPods []corev1.Pod) {
 	})
 }
 
+// handleMasterScaleChange handles Zen updates when a type change results in the addition or the removal of a master:
+// In case of a master scale down it shares the same logic that a "traditional" scale down:
+// * We proactively set m_m_n to the value of 1 if there are 2 Zen1 masters left
+// * We exclude the master for Zen2
+// In case of a master scale up there's nothing else to do:
+// * If there are some Zen1 nodes m_m_n is updated prior the update of the StatefulSet in HandleUpscaleAndSpecChanges
```
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/upgrade_pods_deletion.go]: s/some//

File: pkg/controller/elasticsearch/driver/upgrade_predicates.go [hunk 2]
```diff
@@ -126,21 +135,50 @@ var predicates = [...]Predicate{
 			deletedPods []corev1.Pod,
 			maxUnavailableReached bool,
 		) (b bool, e error) {
-			// If candidate is not a master then we don't care
+
+			// If candidate is not a master then we just check if it will become a master
+			// In this case we account for a master creation as we want to avoid creating more
+			// than one master at a time.
 			if !label.IsMasterNode(candidate) {
+				if stringsutil.StringInSlice(candidate.Name, context.masterNodesNames) {
```
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/upgrade_predicates.go]: Maybe worth putting this behind a speaking function name like `func willBecomeMasterNode` or something like that

File: pkg/controller/elasticsearch/driver/upgrade_predicates_test.go [hunk 3]
```diff
@@ -9,11 +9,157 @@ import (
...
+			wantErr:                      false,
+			wantShardsAllocationDisabled: true,
+		},
+		{
+			name: "Two data nodes converted into master+data nodes, step 2: upgrade the remaining one",
+			fields: fields{
+				upgradeTestPods: newUpgradeTestPods(
+					newTestPod("masters-0").withVersion("7.2.0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("data-to-masters-0").withVersion("7.2.0").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("data-to-masters-1").withVersion("7.2.0").isMaster(false).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
```
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/upgrade_predicates_test.go]: Should this not be `isMaster(true` at this point?

File: pkg/controller/elasticsearch/sset/fixtures.go [hunk 4]
```diff
@@ -23,6 +24,27 @@ type TestSset struct {
...
+				Labels: map[string]string{
+					label.VersionLabelName:     t.Version,
+					label.ClusterNameLabelName: t.ClusterName,
+				},
+			},
+		}
+		label.NodeTypesMasterLabelName.Set(t.Master, pod.Labels)
+		label.NodeTypesDataLabelName.Set(t.Data, pod.Labels)
+		pods[i] = pod
+	}
```
Comment (pebrc) [inline pkg/controller/elasticsearch/sset/fixtures.go]: Could this not reuse `TestPod`?

File: pkg/controller/elasticsearch/version/zen1/compatibility.go [hunk 5]
```diff
@@ -21,8 +24,37 @@ func IsCompatibleWithZen1(statefulSet appsv1.StatefulSet) bool {
 	return sset.ESVersionMatch(statefulSet, versionCompatibleWithZen1)
 }
 
-// AtLeastOneNodeCompatibleWithZen1 returns true if the given StatefulSetList contains
-// at least one StatefulSet compatible with zen1.
-func AtLeastOneNodeCompatibleWithZen1(statefulSets sset.StatefulSetList) bool {
-	return sset.AtLeastOneESVersionMatch(statefulSets, versionCompatibleWithZen1)
+// AtLeastOneNodeCompatibleWithZen1 returns true if at least one of the following condition is true:
```
Comment (pebrc) [inline pkg/controller/elasticsearch/version/zen1/compatibility.go]: conditions

File: pkg/controller/elasticsearch/version/zen2/voting_exclusions.go [hunk 6]
```diff
@@ -45,6 +45,7 @@ func canClearVotingConfigExclusions(c k8s.Client, actualStatefulSets sset.Statef
 	// - nodes are expected to be in the cluster (shouldn't be removed anymore)
 	// They cannot be cleared when:
 	// - expected nodes to remove are not removed yet
+	// - expectation like Pod being restarted should be check prior calling this function
```
Comment (pebrc) [inline pkg/controller/elasticsearch/version/zen2/voting_exclusions.go]: _checked_ prior _to_ calling this function

File: pkg/controller/elasticsearch/sset/fixtures.go (hunk 4 above)
Comment (barkbay) [reply pkg/controller/elasticsearch/sset/fixtures.go]: `TestPod` seems suitable when creating Pods "from scratch"
In my case I can generate them automatically from the `TestSset` without duplicating all the needed information.

**Comment by barkbay** [comment]
jenkins test this please

**Pull request merged**